Tests/Common/Data/PeriodCountConsolidatorTests.cs

```diff
@@ -115,5 +115,160 @@ public void ExpectedConsolidatedQuoteBarsInPeriodMode(TimeSpan barSpan, TimeSpan
+            // But the emit time is slightly offset (the timeslice that contains the bar)
+            // So add a random ms offset to the scan time
+            consolidator.Update(new TradeBar { Time = time, Period = Time.OneHour });
+            time = time.Add(period);
+            consolidator.Scan(time.AddMilliseconds(random.Next(800))); 
+
+            consolidator.Update(new TradeBar { Time = time, Period = Time.OneHour });
+            time = time.Add(period);
+            consolidator.Scan(time.AddMilliseconds(random.Next(800)));
+
+            consolidator.Update(new TradeBar { Time = time, Period = Time.OneHour });
+            time = time.Add(period);
+            consolidator.Scan(time.AddMilliseconds(random.Next(800)));
+
+            consolidator.Update(new TradeBar { Time = time, Period = Time.OneHour });
+            time = time.Add(period);
+            consolidator.Scan(time.AddMilliseconds(random.Next(800)));
+
+            // We should expect to see 4 bars emitted from the consolidator
+            Assert.AreEqual(4,consolidatedBarsCount);
+        }
+
+        [Test]
+        public void ConsolidatorEmitsOldBarsUsingUpdate()
+        {
+            // This test is to ensure that no bars get swallowed by the consolidator
+            // even if it doesn't get the data on regular intervals.
+            // We will use the PushThrough method which calls update
+            var period = TimeSpan.FromHours(1);
+            var consolidator = new TradeBarConsolidator(period);
+            TradeBar latestConsolidated = null;
+            var consolidatedBarsCount = 0;
+
+            consolidator.DataConsolidated += (sender, bar) =>
+            {
+                latestConsolidated = bar;
+                consolidatedBarsCount++;
+            };
+
+            // Set our starting time 04/13/2015 at 12:00AM
+            var time = new DateTime(2015, 04, 13);
+
+            // Update this consolidator with minute tradebars but one less than 60, which would trigger emit
+            PushBarsThrough( 59, Time.OneMinute, consolidator, ref time);
+
+            // No bars should be emitted, lets assert the current time and count
+            Assert.IsTrue(time == new DateTime(2015, 04, 13, 0, 59, 0));
+            Assert.AreEqual(0, consolidatedBarsCount);
+
+            // Advance time way past (3 hours) the bar end time of 1AM
+            time += TimeSpan.FromHours(3); // Time = 3:59AM now
+
+            // Push one bar through at 3:59AM and check that we still get the 12AM - 1AM Bar emitted
+            PushBarsThrough(1, Time.OneMinute, consolidator, ref time);
+            Assert.AreEqual(1, consolidatedBarsCount);
+            Assert.IsTrue(latestConsolidated != null && latestConsolidated.Time == new DateTime(2015, 04, 13));
+
+            // Check the new working bar is 3AM to 4AM, This is because we pushed a bar in at 3:59AM
+            Assert.IsTrue(consolidator.WorkingBar.Time == new DateTime(2015, 04, 13, 3, 0, 0));
```

Comment:
Please add another unit test to literally cover the 2pm, 3pm, 4pm case for hourly bars. 95%+ of algorithms won't get data past 4pm until the following morning, but it'll seem like a bug if they don't get the 4pm bar until the next day.